src/Http/Routing/src/Matching/CandidateSet.cs

```diff
@@ -281,6 +281,57 @@ public void SetValidity(int index, bool value)
             }
         }
 
+        /// <summary>
+        /// Replaces the <see cref="Endpoint"/> at the provided <paramref name="index"/> with the
+        /// provided <paramref name="endpoint"/>.
+        /// </summary>
+        /// <param name="index">The candidate index.</param>
+        /// <param name="endpoint">
+        /// The <see cref="Endpoint"/> to replace the original <see cref="Endpoint"/> at
+        /// the <paramref name="index"/>. If <paramref name="endpoint"/> the candidate will be marked
+        /// as invalid.
+        /// </param>
+        /// <param name="values">
+        /// The <see cref="RouteValueDictionary"/> to replace the original <see cref="RouteValueDictionary"/> at
+        /// the <paramref name="index"/>.
+        /// </param>
+        public void ReplaceEndpoint(int index, Endpoint endpoint, RouteValueDictionary values)
```

Comment:
Not familiar with the code, but are there any concurrency concerns here if this is invoked by multiple invocations of matcher policy?

Reply:
`CandidateSet` is per-request mutable state